File: Audio/Visualization/VisualizationController.m
//
//  VisualizationController.m
//  CogAudio Framework
//
//  Created by Christopher Snowhill on 2/12/22.
//

#import <CogAudio/VisualizationController.h>

#import <Accelerate/Accelerate.h>

#import "fft.h"

@implementation VisualizationController {
	MIDIVisualizationController *midiController;
	NSTimer *sineTimer;

	double sampleRate;
	double latency;
	double fullLatency;
	float *visAudio;
	int visAudioCursor, visAudioSize;
	uint64_t visSamplesPosted;
	BOOL ignoreLatency;
	double sinePhase;
}

+ (VisualizationController *)sharedController {
	static VisualizationController *_sharedController = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		_sharedController = [[VisualizationController alloc] init];
	});
	return _sharedController;
}

+ (MIDIVisualizationController *)sharedMIDIController {
	return [[VisualizationController sharedController] midiController];
}

- (MIDIVisualizationController *)midiController {
	return midiController;
}

- (id)init {
	self = [super init];
	if(self) {
		midiController = [[MIDIVisualizationController alloc] initWithController:self];

		visAudio = NULL;
		visAudioSize = 0;
		latency = 0;
		ignoreLatency = YES;
		sinePhase = 0.0;

		sineTimer = [NSTimer timerWithTimeInterval:1.0 / 60.0 repeats:YES block:^(NSTimer * _Nonnull timer) {
			double sinePhase = self->sinePhase;
			self->sinePhase = fmod(sinePhase + (M_PI / 90.0), M_PI * 2.0);
		}];
		[[NSRunLoop mainRunLoop] addTimer:sineTimer forMode:NSRunLoopCommonModes];
	}
	return self;
}

- (void)dealloc {
	fft_free();
	[sineTimer invalidate];
	sineTimer = nil;
}

- (void)reset {
	@synchronized (self) {
		latency = 0;
		visAudioCursor = 0;
		visSamplesPosted = 0;
		ignoreLatency = YES;
		sinePhase = 0.0;
		if(visAudio && visAudioSize) {
			bzero(visAudio, sizeof(float) * visAudioSize);
		}
	}
}

- (void)postSampleRate:(double)sampleRate {
	@synchronized(self) {
		if(self->sampleRate != sampleRate) {
			self->sampleRate = sampleRate;
			int visAudioSize = (int)(sampleRate * 45.0);
			void *visAudio = realloc(self->visAudio, visAudioSize * sizeof(float));
			if(visAudio && visAudioSize) {
				if(visAudioSize > self->visAudioSize) {
					bzero(((float *)visAudio) + self->visAudioSize, sizeof(float) * (visAudioSize - self->visAudioSize));
				}
				self->visAudio = visAudio;
				self->visAudioSize = visAudioSize;
				visAudioCursor %= visAudioSize;
			} else {
				if(self->visAudio) {
					free(self->visAudio);
					self->visAudio = NULL;
				}
				self->visAudioSize = 0;
			}
		}
	}
}

- (void)postVisPCM:(const float *)inPCM amount:(int)amount {
	@synchronized(self) {
		if(!visAudioSize) {
			return;
		}
		int samplesRead = 0;
		while(amount > 0) {
			int amountToCopy = (int)(visAudioSize - visAudioCursor);
			if(amountToCopy > amount) amountToCopy = amount;
			cblas_scopy(amountToCopy, inPCM + samplesRead, 1, visAudio + visAudioCursor, 1);
			visAudioCursor = visAudioCursor + amountToCopy;
			if(visAudioCursor >= visAudioSize) visAudioCursor -= visAudioSize;
			amount -= amountToCopy;
			samplesRead += amountToCopy;
			visSamplesPosted += amountToCopy;
		}
	}
}

- (void)postLatency:(double)latency {
	if((latency >= 45.0) || (latency < 0.0)) [[clang::unlikely]] {
		ignoreLatency = YES;
		self->latency = latency;
	} else {
		self->latency = latency;
		ignoreLatency = NO;
	}
}

- (double)getLatency {
	if(!ignoreLatency) return latency;
	else return 0.0;
}

- (void)postFullLatency:(double)latency {
	self->fullLatency = latency;
}

- (double)getFullLatency {
	if(!ignoreLatency) return fullLatency;
	else return 0.0;
}

- (double)readSampleRate {
	@synchronized(self) {
		return sampleRate ?: 44100.0;
	}
}

- (UInt64)samplesPosted {
	return visSamplesPosted;
}

- (void)generateSineWave:(float *_Nullable)outPCM visFFT:(float *_Nullable)outFFT {
	double sinePhase = self->sinePhase;
	if(outPCM || outFFT) {
		const double stepSize = M_PI * 2.0 * 5.0 / 4096.0;
		double sineStep = sinePhase;
		for(int i = 0; i < 2048; ++i) {
			double sinePoint = sin(sineStep);
			if(outPCM) {
				outPCM[i * 2] = sinePoint;
				outPCM[i * 2 + 1] = sin(sineStep + stepSize);
			}
			if(outFFT) {
				outFFT[i] = sinePoint * -40.0 - 40.0;
			}
			sineStep = fmod(sineStep + stepSize * 2, M_PI * 2.0);
		}
	}
}

- (void)copyVisPCM:(float *_Nullable)outPCM visFFT:(float *_Nullable)outFFT latencyOffset:(double)latency {
	if(!outPCM && !outFFT) return;

	if(ignoreLatency || !visAudio || !visAudioSize) {
		[self generateSineWave:outPCM visFFT:outFFT];
		return;
	}

	void *visAudioTemp = calloc(sizeof(float), 4096);
	if(!visAudioTemp) {
		[self generateSineWave:outPCM visFFT:outFFT];
		return;
	}

	@synchronized(self) {
		if(!sampleRate) {
			free(visAudioTemp);
			[self generateSineWave:outPCM visFFT:outFFT];
			return;
		}
		int latencySamples = (int)(sampleRate * (self->latency + latency)) + 2048;
		if(latencySamples < 4096) latencySamples = 4096;
		int readCursor = visAudioCursor - latencySamples;
		int samples = 4096;
		int samplesRead = 0;
		if(latencySamples + samples > visAudioSize) {
			samples = (int)(visAudioSize - latencySamples);
		}
		while(readCursor < 0)
			readCursor += visAudioSize;
		while(readCursor >= visAudioSize)
			readCursor -= visAudioSize;
		while(samples > 0) {
			int samplesToRead = (int)(visAudioSize - readCursor);
			if(samplesToRead > samples) samplesToRead = samples;
			cblas_scopy(samplesToRead, visAudio + readCursor, 1, visAudioTemp + samplesRead, 1);
			samplesRead += samplesToRead;
			readCursor += samplesToRead;
			samples -= samplesToRead;
			if(readCursor >= visAudioSize) readCursor -= visAudioSize;
		}
	}
	if(outPCM) {
		cblas_scopy(4096, visAudioTemp, 1, outPCM, 1);
	}
	if(outFFT) {
		fft_calculate(visAudioTemp, outFFT, 2048);
	}

	free(visAudioTemp);
}

@end

File: Audio/Visualization/VisualizationController.swift
//
//  VisualizationController.swift
//  CogAudio Framework
//
//  Created by Christopher Snowhill on 6/30/22.
//

import Foundation

@objc(VisualizationController)
class VisualizationController : NSObject {
	var serialQueue = DispatchQueue(label: "Visualization Queue")
	var sampleRate = 0.0
	var latency = 0.0
	var visAudio: [Float] = Array(repeating: 0.0, count: 44100 * 45)
	var visAudioCursor = 0
	var visAudioSize = 0
	var visSamplesPosted: UInt64 = 0

	private static var sharedVisualizationController: VisualizationController = {
		let visualizationController = VisualizationController()
		return visualizationController
	}()

	@objc
	class func sharedController() -> VisualizationController {
		return sharedVisualizationController
	}
	
	@objc
	func reset() {
		serialQueue.sync {
			self.latency = 0;
			let amount = self.visAudioSize
			for i in 0..<amount {
				self.visAudio[i] = 0
			}
			self.visSamplesPosted = 0;
		}
	}

	@objc
	func postLatency(_ latency: Double) {
		self.latency = latency
	}
	
	@objc
	func samplesPosted() -> UInt64 {
		return self.visSamplesPosted
	}

	@objc
	func postSampleRate(_ sampleRate: Double) {
		serialQueue.sync {
			if(self.sampleRate != sampleRate) {
				self.sampleRate = sampleRate
				visAudioSize = (Int)(sampleRate * 45.0)
				visAudio = Array(repeating: 0.0, count: visAudioSize)
				visAudioCursor = 0
			}
		}
	}

	@objc
	func postVisPCM(_ inPCM: UnsafePointer<Float>?, amount: Int) {
		serialQueue.sync {
			let bufferPointer = UnsafeBufferPointer<Float>(start: inPCM, count: amount)
			var j = self.visAudioCursor
			let k = self.visAudioSize
			if(j + amount <= k) {
				let endIndex = j + amount;
				self.visAudio.replaceSubrange(j..<endIndex, with: bufferPointer)
				j += amount
				if(j >= k) { j = 0 }
			} else {
				let inEndIndex = k - j
				let remainder = amount - inEndIndex
				self.visAudio.replaceSubrange(j..<k, with: bufferPointer.prefix(inEndIndex))
				self.visAudio.replaceSubrange(0..<remainder, with: bufferPointer.suffix(remainder))
				j = remainder
			}
			self.visAudioCursor = j
			self.visSamplesPosted += UInt64(amount);
		}
	}

	@objc
	func readSampleRate() -> Double {
		serialQueue.sync {
			return self.sampleRate
		}
	}

	@objc
	func copyVisPCM(_ outPCM: UnsafeMutablePointer<Float>?, visFFT: UnsafeMutablePointer<Float>?, latencyOffset: Double) {
		if(self.visAudioSize == 0) {
			outPCM?.update(repeating: 0.0, count: 4096)
			visFFT?.update(repeating: 0.0, count: 2048)
			return
		}

		var outPCMCopy = Array<Float>(repeating: 0.0, count: 4096)

		serialQueue.sync {
			// Offset latency so the target sample is in the center of the window
			let latencySamples = (Int)((self.latency + latencyOffset) * self.sampleRate) + 2048
			var samplesToDo = 4096;
			if(latencySamples < 0) {
				return;
			}
			if(latencySamples < 4096) {
				// Latency can sometimes dip below this threshold
				samplesToDo = latencySamples;
			}
			var j = self.visAudioCursor - latencySamples
			let k = self.visAudioSize
			if j < 0 { j += k }
			if(j + samplesToDo <= k) {
				outPCMCopy.replaceSubrange(0..<samplesToDo, with: self.visAudio.suffix(from: j).prefix(samplesToDo))
			} else {
				let outEndIndex = k - j
				let remainder = samplesToDo - outEndIndex
				outPCMCopy.replaceSubrange(0..<outEndIndex, with: self.visAudio.suffix(from: j))
				outPCMCopy.replaceSubrange(outEndIndex..<samplesToDo, with: self.visAudio.prefix(remainder))
			}
		}

		outPCM?.update(from: outPCMCopy, count: 4096)

		if(visFFT != nil) {
			serialQueue.sync {
				fft_calculate(outPCMCopy, visFFT, 2048)
			}
		}
	}
}
